x/dex/module.go

```diff
@@ -246,25 +245,93 @@ func (am AppModule) beginBlockForContract(ctx sdk.Context, contractAddr string)
 // EndBlock executes all ABCI EndBlock logic respective to the capability module. It
 // returns no validator updates.
 func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
-	for _, contractAddr := range am.getAllContractAddresses(ctx) {
-		ctx.Logger().Info(fmt.Sprintf("End block for %s", contractAddr))
-		am.endBlockForContract(ctx, contractAddr)
+	validContractAddresses := map[string]types.ContractInfo{}
+	for _, contractInfo := range am.getAllContractInfo(ctx) {
+		validContractAddresses[contractInfo.ContractAddr] = contractInfo
+	}
+	// Each iteration is atomic. If an iteration finishes without any error, it will return,
+	// otherwise it will rollback any state change, filter out contracts that cause the error,
+	// and proceed to the next iteration. The loop is guaranteed to finish since
+	// `validContractAddresses` will always decrease in size every iteration.
+	for len(validContractAddresses) > 0 {
+		failedContractAddresses := utils.NewStringSet([]string{})
+		cachedCtx, msCached := store.GetCachedContext(ctx)
+		// cache keeper in-memory state
+		memStateCopy := am.keeper.MemState.DeepCopy()
+		finalizeBlockMessages := map[string]*types.SudoFinalizeBlockMsg{}
+		for contractAddr := range validContractAddresses {
+			finalizeBlockMessages[contractAddr] = types.NewSudoFinalizeBlockMsg()
+		}
+
+		for contractAddr, contractInfo := range validContractAddresses {
+			if contractInfo.HookOnly {
```

Comment:
the variable name hookOnly may be a little bit confusing, cuz every contract will be able to return a hook (callback), but not each of them needs to process limit/market order. So maybe it's better to create two variables, one as `NeedCallBack` or ` NeedHook`, another one as `NeedOrderMatching` (default to true)